mjolnir/input/read_local_interaction.hpp

```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+
+    return make_unique<DirectionalContactInteraction<traitsT, angle1_potentialT,
+               angle2_potentialT, contact_potentialT>>(
+                   kind, read_directional_contact_potentials<
+                   4, real_type, angle1_potentialT,
+                   angle2_potentialT, contact_potentialT>(local));
+  }
+  else if(contact_potential == "Gaussian")
+  {
+    MJOLNIR_LOG_NOTICE("-- contact potential function is Gaussian.");
+    using contact_potentialT = GaussianPotential<real_type>;
+
+    return make_unique<DirectionalContactInteraction<traitsT, angle1_potentialT,
+               angle2_potentialT, contact_potentialT>>(
+                   kind, read_directional_contact_potentials<
+                   4, real_type, angle1_potentialT,
+                   angle2_potentialT, contact_potentialT>(local));
+  }
+  else if(contact_potential == "Uniform")
+  {
+    MJOLNIR_LOG_NOTICE("-- contact potential function is Uniform potential");
+    using contact_potentialT = UniformPotential<real_type>;
+
+    return make_unique<DirectionalContactInteraction<traitsT, angle1_potentialT,
+               angle2_potentialT, contact_potentialT>>(
+                   kind, read_directional_contact_potentials<
+                   4, real_type, angle1_potentialT,
+                   angle2_potentialT, contact_potentialT>(local));
+  }
+  else
+  {
+    throw_exception<std::runtime_error>(toml::format_error("[error] "
+        "mjolnir::read_directional_contact_interaction: invalid contact potential",
+        toml::find<toml::value>(local, "potentials", "contact"), "here", {
+        "expected value is one of the following.",
+        "- \"GoContact\": r^12 - r^10 type native contact potential",
+        "- \"Gaussian\" : well-known gaussian potential"
+        }));
+  }
+}
+
+// This is internal function for read_directional_contact_interaction function.
+template<typename traitsT, typename ... PotentialTs>
+typename std::enable_if<sizeof...(PotentialTs) == 2,
+                        std::unique_ptr<LocalInteractionBase<traitsT>>>::type
+read_angle_in_directional_contact_interaction(
+    const std::string& kind, const toml::value& local, std::vector<std::string>&)
+{
+  return read_contact_in_directional_contact_interaction<traitsT, PotentialTs...>(kind, local);
+}
+
+// This is internal function for read_directional_contact_interaction function.
+template<typename traitsT, typename ... PotentialTs>
+typename std::enable_if<sizeof...(PotentialTs) < 2,
+                        std::unique_ptr<LocalInteractionBase<traitsT>>>::type
+read_angle_in_directional_contact_interaction(
+    const std::string& kind, const toml::value& local, std::vector<std::string>& angle_potential_keys)
+{
+  MJOLNIR_GET_DEFAULT_LOGGER();
+  using real_type = typename traitsT::real_type;
+
+  std::string angle_potential_key = angle_potential_keys.back();
+  std::string angle_potential = toml::find<std::string>(local, "potentials", angle_potential_key);
```

Comment:
Here, the types are enough clear. You can use `auto`.
Also, you must use `const` because these values will never be modified. You need to avoid accidental modification on those variables.
```suggestion
  const auto angle_potential = toml::find<std::string>(local, "potentials", angle_potential_key);
```